pkg/applicationserver/io/packages/alcsync/v1/observability.go

```diff
@@ -0,0 +1,119 @@
+// Copyright © 2023 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package alcsyncv1
+
+import (
+	"context"
+	"fmt"
+
+	"go.thethings.network/lorawan-stack/v3/pkg/events"
+	"go.thethings.network/lorawan-stack/v3/pkg/log"
+	"go.thethings.network/lorawan-stack/v3/pkg/ttnpb"
+)
+
+func publishEvents(ctx context.Context, builders ...events.Builder) {
+	n := len(builders)
+	if n == 0 {
+		return
+	}
+
+	evts := make([]events.Event, n)
+	for i, builder := range builders {
+		evts[i] = builder.New(ctx)
+	}
+	log.FromContext(ctx).WithField("event_count", n).Debug("Publish events")
+	events.Publish(evts...)
+}
+
+func alcsyncEventOptions(extraOpts ...events.Option) []events.Option {
+	return append([]events.Option{events.WithVisibility(ttnpb.Right_RIGHT_APPLICATION_TRAFFIC_READ)}, extraOpts...)
+}
+
+func defineALCSyncPkgFailEvent(name, desc string, opts ...events.Option) func() events.Builder {
+	return events.DefineFunc(
+		fmt.Sprintf("as.packages.alcsync.%s.failed", name), fmt.Sprintf("%s package failed", desc),
+		alcsyncEventOptions(opts...)...,
+	)
+}
+
+func defineALCSyncCmdReceivedEvent(name, desc string, opts ...events.Option) func() events.Builder {
+	return events.DefineFunc(
+		fmt.Sprintf("as.packages.alcsync.cmd.%s.accept", name), fmt.Sprintf("%s command received", desc),
+		alcsyncEventOptions(opts...)...,
+	)
+}
+
+func defineALCSyncCmdParsedEvent(name, desc string, opts ...events.Option) func() events.Builder {
+	return events.DefineFunc(
+		fmt.Sprintf("as.packages.alcsync.cmd.%s.parsed", name), fmt.Sprintf("%s command parsed", desc),
+		alcsyncEventOptions(opts...)...,
+	)
+}
+
+func defineALCSyncCmdParsedFailEvent(name, desc string, opts ...events.Option) func() events.Builder {
```

Comment:
The thing about parsing bytes is that it is actually hard to check if the command that failed was even _real_. 

Think of it like this - due to some mismatch in standards / misbehavior a command is now extended from one byte of payload to two bytes of payload. You parse the first byte, and you consumption function now uses the second byte of payload as a command identifier - if you are lucky, it's an invalid identifier, but if you are not, it will just look like this fictitious command has an invalid payload, while it was not there in the first place.

As such, I would recommend that we just have a generic parse failed event, not one per request type.